File: core/src/main/java/com/palomiklo/streampulse/listener/Listener.java
package com.palomiklo.streampulse.listener;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.palomiklo.streampulse.connection.ConnectionHolder;

import jakarta.servlet.AsyncContext;
import jakarta.servlet.AsyncEvent;
import jakarta.servlet.AsyncListener;

public class Listener {

    private static final Logger log = LoggerFactory.getLogger(Listener.class);

    public static void addListeners(AsyncContext actx, ConnectionHolder chldr) {
        actx.addListener(new AsyncListener() {
            @Override
            public void onComplete(AsyncEvent event) {
                log.debug("Async process completed for connection ID: {}", chldr.id());
                if (chldr.connection().isConnected()) {
                    chldr.connection().closeConnection();
                }
            }

            @Override
            public void onTimeout(AsyncEvent event) {
                log.debug("Async process timed out for connection ID: {}", chldr.id());
                if (chldr.connection().isConnected()) {
                    chldr.connection().closeConnection();
                }
            }

            @Override
            public void onError(AsyncEvent event) {
                log.error("Error occurred for connection ID: {}: {}", chldr.id(), event.getThrowable().getLocalizedMessage());
                actx.complete();
                if (chldr.connection().isConnected()) {
                    chldr.connection().closeConnection();
                }
            }

            @Override
            public void onStartAsync(AsyncEvent event) {
                log.debug("Async process started for connection ID: {}", chldr.id());
            }
        });
    }

}

File: core/src/main/kotlin/com/palomiklo/streampulse/listener/Listener.kt
package com.palomiklo.streampulse.listener

import com.palomiklo.streampulse.connection.ConnectionHolder
import jakarta.servlet.AsyncContext
import jakarta.servlet.AsyncEvent
import jakarta.servlet.AsyncListener
import org.slf4j.Logger
import org.slf4j.LoggerFactory

object Listener {
    private val log: Logger = LoggerFactory.getLogger(Listener::class.java)

    fun addListeners(actx: AsyncContext, chldr: ConnectionHolder) {
        actx.addListener(object : AsyncListener {
            override fun onComplete(event: AsyncEvent) =
                log.debug("Async process completed for connection ID: {}", chldr.connection.id)
                    .also { chldr.connection.takeIf { it.isConnected() }?.closeConnection() }

            override fun onTimeout(event: AsyncEvent) = log.debug("Async process timed out for connection ID: {}", chldr.connection.id)
                .also { chldr.connection.takeIf { it.isConnected() }?.closeConnection() }

            override fun onError(event: AsyncEvent) =
                log.error("Error occurred for connection ID: {}: {}", chldr.connection.id, event.throwable.localizedMessage)
                    .also { actx.complete() }
                    .also { chldr.connection.takeIf { it.isConnected() }?.closeConnection() }

            override fun onStartAsync(event: AsyncEvent) = log.debug("Async process started for connection ID: {}", chldr.connection.id)
        })
    }
}
